tools/internal/parser/file.go

```diff
@@ -0,0 +1,204 @@
+	// warnings. Warnings are a concession to old PSL entries that now
+	// have validation errors, due to PSL policy changes. As long as
+	// the entries in question don't change, their preexisting
+	// validation errors are downgraded to lint warnings.
+	Warnings []error
+}
+
+// AllSuffixBlocks returns all suffix blocks in f.
+func (f *File) AllSuffixBlocks() []Suffixes {
+	var ret []Suffixes
+
+	for _, block := range f.Blocks {
+		switch v := block.(type) {
+		case Suffixes:
+			ret = append(ret, v)
+		}
+	}
+
+	return ret
+}
+
+// SuffixBlocksInSection returns all suffix blocks within the named
+// file section (for example, "ICANN DOMAINS" or "PRIVATE DOMAINS").
+func (f *File) SuffixBlocksInSection(name string) []Suffixes {
+	var ret []Suffixes
+
+	var curSection string
+	for _, block := range f.Blocks {
+		switch v := block.(type) {
+		case StartSection:
+			curSection = v.Name
+		case EndSection:
+			if curSection == name {
+				return ret
+			}
+			curSection = ""
+		case Suffixes:
+			if curSection == name {
+				ret = append(ret, v)
+			}
+		}
+	}
+	return ret
+}
+
+// Source is a piece of source text with location information.
+type Source struct {
+	// StartLine is the first line of this piece of source text in the
+	// original file. The first line of a file is line 1 rather than
+	// line 0, since that is how text editors conventionally number
+	// lines.
+	StartLine int
+	// EndLine is the last line of this piece of source text in the
+	// original file. The line named by EndLine is included in the
+	// source block.
+	EndLine int
+	// Raw is the unparsed source text for this block.
+	Raw string
+}
+
+// LocationString returns a short string describing the source
+// location.
+func (s Source) LocationString() string {
+	if s.StartLine == s.EndLine {
+		return fmt.Sprintf("line %d", s.StartLine)
+	}
+	return fmt.Sprintf("lines %d-%d", s.StartLine, s.EndLine)
+}
+
+// A Block is a parsed chunk of a PSL file.
+// In Parse's output, a Block is one of the following concrete types:
+// Comment, StartSection, EndSection, Suffixes.
+type Block interface {
+	source() Source
+}
+
+// Comment is a standalone top-level comment block.
+type Comment struct {
+	Source
+}
+
+func (c Comment) source() Source { return c.Source }
+
+// StartSection is a top-level marker that indicates the start of a
+// logical section, such as ICANN suffixes or privately managed
+// domains.
+//
+// Sections cannot be nested, at any one point in a file you are
+// either not in any logical section, or within a single section.  In
+// a File that has no parse errors, StartSection and EndSection blocks
+// are correctly paired, and all sections are closed by an EndSection
+// before any following StartSection.
+type StartSection struct {
+	Source
+	Name string // section name, e.g. "ICANN DOMAINS", "PRIVATE DOMAINS"
+}
+
+func (b StartSection) source() Source { return b.Source }
+
+// EndSection is a top-level marker that indicates the end of a
```

Comment:
Maybe ditch the duplicated comment and make the one above talk about Start & End section? Unless this is because of docstrings/structured comments.

Reply:
Yeah, it's because of docstrings, the Go package documentation takes the comment attached to the struct when it's generating the HTML docs. When I wrote this I checked the look of the generated docs, and unfortunately duplication makes things easier to understand :/